File: apps/fastapi-backend/ai_backends/langgraph.py
"""LangGraph AI backend — exposes two topologies:

  - "react"        → `langgraph.prebuilt.create_react_agent`. Tool-calling loop,
                     prebuilt convenience function. The default and most common
                     LangGraph pattern.
  - "plan-execute" → custom StateGraph (planner → executor → replanner) ported
                     from the official notebook
                     https://github.com/langchain-ai/langgraph/blob/main/examples/plan-and-execute/plan-and-execute.ipynb
                     Showcases LangGraph's actual abstraction beyond the
                     prebuilt — separate phases, explicit nodes, conditional
                     edges.

Both topologies emit raw `astream_events` JSON over the wire — same format that
`langgraph-cli dev` and LangGraph Cloud produce. The matching `langGraphAdapter`
in packages/server/src/adapters/langgraph.ts normalizes either to AI SDK v6.
"""

import json
import operator
from typing import Annotated, List, Tuple, TypedDict, Union

from langchain_core.prompts import ChatPromptTemplate
from langgraph.checkpoint.memory import InMemorySaver
from langgraph.graph import END, START, StateGraph
from langgraph.prebuilt import create_react_agent
from langgraph.types import interrupt
from pydantic import BaseModel, Field

from ._common import (
    _DECISION_TYPES,
    _pending_approval_events,
    approval_interrupt_on,
    approval_resume_command,
    approval_thread_config,
    SYSTEM_PROMPT,
    TOOLS,
    langfuse_config,
    make_llm,
)


_INTERESTING_EVENTS = frozenset(
    {"on_chat_model_stream", "on_tool_start", "on_tool_end"}
)

# Nodes whose `on_chat_model_stream` events emit raw JSON from
# `with_structured_output` chains. We hide these tokens from the wire — they
# are internal data-shape generation, not user-facing prose. Tool calls and
# the final response surface through other channels (the executor node's
# natural-language stream and the StateGraph state's `response` field).
_STRUCTURED_OUTPUT_NODES = frozenset({"planner", "replanner"})


def _should_emit(event: dict) -> bool:
    if event.get("event") not in _INTERESTING_EVENTS:
        return False
    if event.get("event") == "on_chat_model_stream":
        node = (event.get("metadata") or {}).get("langgraph_node", "")
        if node in _STRUCTURED_OUTPUT_NODES:
            return False
    return True


# ---------------------------------------------------------------------------
# Topology 1: ReAct (prebuilt)
# ---------------------------------------------------------------------------

_react_graph = None


def get_react_graph():
    global _react_graph
    if _react_graph is None:
        _react_graph = create_react_agent(
            make_llm(),
            tools=TOOLS,
            prompt=SYSTEM_PROMPT,
            name="fastapi-langgraph-react",
        )
    return _react_graph


# Kept for backward compat with main.py lifespan and any callers that imported
# `langgraph.get_graph()` before the topology axis existed.
def get_graph():
    return get_react_graph()


# ---------------------------------------------------------------------------
# THE APPROVAL GATE FOR THIS RUNG (#332 step C2/C3).
#
# WHY NOT `interrupt_before=["tools"]`, WHICH IS WHAT #332's TABLE SAYS. Measured
# against langgraph 1.2.11 before choosing, and it fails on two counts:
#
#   * IT EMITS NO PAYLOAD. `interrupt_before` stops the graph before a node; it
#     does not call `interrupt()`, so `state.tasks[].interrupts` is EMPTY. The
#     effect is correctly withheld -- 0 side effects at the pause, measured --
#     and the client is told nothing: a 200, one empty message frame, silence.
#     That is the exact defect langchain.py documents as the reason the gate
#     could not be armed until an approval frame existed (#413 shipped disarmed
#     for it), and it would have been reintroduced here by following the table.
#
#   * IT IS NODE-LEVEL, AND THE POLICY IS PER-TOOL. `interrupt_before` pauses
#     before the whole tools node whatever tool was called, so a request whose
#     policy allowlists `get_counter` as read-only would still pause on it. The
#     other plane, using HumanInTheLoopMiddleware's `interrupt_on` map, would
#     not. Two runtimes answering the same policy differently is precisely the
#     divergence this repo exists to make visible.
#
# So the gate is langgraph's own `interrupt()`, called from a `post_model_hook`
# that reads the model's proposed tool calls and pauses only on the ones the
# request's policy does not excuse. Measured, four cells:
#
#   gated tool       0 effects at the pause, 1 interrupt, effect runs on resume
#   allowlisted tool no pause, no interrupt, and the tool RAN -- the presence
#                    companion, without which "did not pause" is also satisfied
#                    by a tool that never executed at all
#
# WE AUTHOR THIS PAYLOAD, AND THAT IS A REAL COST. #332 says the four-way
# vocabulary is carried faithfully rather than translated, which on the
# langchain rung means passing upstream's own dict through. There is no upstream
# payload to pass here: the middleware that builds one lives in langchain, and
# reaching for it would make this rung a copy of that one rather than a LangGraph
# demonstration. So the shape is authored to match, and
# `test_approval_payload_shape` asserts it against what the langchain plane
# actually emits rather than against a copy of this dict -- otherwise the two
# drift and nothing compares them.
# ---------------------------------------------------------------------------

# ONE SAVER FOR THE PROCESS, NOT ONE PER REQUEST -- the reasoning is
# langchain.py's and is the same here: a decision arrives on a LATER request,
# and a per-request saver makes every approval the lost-checkpoint case.
_APPROVAL_SAVER = InMemorySaver()


def _approval_gate(state):
    """Pause before any tool call this request's policy does not excuse.

    Runs after the model and before the tools node, which is the only point
    where the proposed calls are known and none of them has run yet.
    """
    last = state["messages"][-1]
    interrupt_on = approval_interrupt_on(t.name for t in TOOLS)
    pending = [
        call
        for call in getattr(last, "tool_calls", None) or []
        if interrupt_on.get(call["name"])
    ]
    if not pending:
        # NOT AN ERROR AND NOT A GAP. Either the model called nothing, or every
        # call it made is allowlisted -- and an allowlisted call is meant to run.
        return None
    # THE SHAPE UPSTREAM ACTUALLY EMITS, MEASURED RATHER THAN QUOTED.
    #
    # `action_requests` paired BY INDEX with `review_configs`, snake_case, which
    # is what docs/sse-frame-schema.json calls the contract and what
    # packages/react's ApprovalPauseSchema parses. Measured on langchain 1.3.18
    # and deepagents 0.7.11: both middlewares emit
    #
    #     action_requests: [{name, args, description}]
    #     review_configs:  [{action_name, allowed_decisions}]
    #
    # #332's issue body quotes a flat `{action_name, allowed_decisions}`, which
    # neither produces. The first version of this function was built from that
    # quote and emitted `action_requests[].action_name` with a top-level
    # `allowed_decisions` — a payload the card's schema rejects, so a gated pause
    # would have rendered with no action name and no buttons. The rung's own test
    # asserted the invented shape and was green.
    #
    # THIS RUNG AUTHORS THE PAYLOAD and the other two pass upstream's through, so
    # only here can the two drift. packages/test-utils' approval-pause
    # conformance suite drives the real adapter against the real schema for every
    # gating rung, which is what makes the drift visible rather than latent.
    interrupt(
        {
            "action_requests": [
                {
                    "name": call["name"],
                    "args": call["args"],
                    "description": (
                        "Tool execution requires approval\n\n"
                        f"Tool: {call['name']}\nArgs: {call['args']}"
                    ),
                }
                for call in pending
            ],
            # PAIRED BY INDEX with action_requests above, one config per call.
            #
            # THE VOCABULARY WE WILL ACCEPT, NOT A LITERAL BESIDE IT.
            # `_DECISION_TYPES` is what parse_approval_decisions accepts on the
            # way back in; offering a decision the parser would reject is a
            # promise the next request breaks, and two hardcoded lists in one
            # repo is how they come to differ.
            "review_configs": [
                {
                    "action_name": call["name"],
                    "allowed_decisions": list(_DECISION_TYPES),
                }
                for call in pending
            ],
        }
    )
    return None


def get_gated_react_graph():
    """Build this request's react agent, gated by the policy the dispatch parsed.

    Built fresh each call rather than cached in a module global: the policy is
    per-request, so a cached graph would serve the first request's allowlist to
    every later one.
    """
    return create_react_agent(
        make_llm(),
        tools=TOOLS,
        prompt=SYSTEM_PROMPT,
        name="fastapi-langgraph-react",
        post_model_hook=_approval_gate,
        checkpointer=_APPROVAL_SAVER,
    )


# ---------------------------------------------------------------------------
# Topology 2: Plan-and-Execute (StateGraph custom topology)
# ---------------------------------------------------------------------------


class PlanExecuteState(TypedDict):
    """State carried through the plan-execute graph.

    `input`: the original user request — never mutated.
    `plan`: ordered list of remaining steps. Mutated by replanner.
    `past_steps`: append-only log of (step, result) pairs. The Annotated
        operator.add tells LangGraph this field is reduced by concatenation
        when nodes return partial updates.
    `response`: terminal field; non-empty signals "we're done — exit graph."
    """

    input: str
    plan: List[str]
    past_steps: Annotated[List[Tuple[str, str]], operator.add]
    response: str


class _Plan(BaseModel):
    """Structured output schema for the planner."""

    steps: List[str] = Field(
        description="Ordered, atomic steps to accomplish the user's request. "
        "Each step should be specific enough that an executor agent can act "
        "on it in a single turn."
    )


class _Response(BaseModel):
    """Terminal response — used by the replanner when no more steps needed."""

    response: str


class _ReplanAction(BaseModel):
    """Replanner output — either continue with new plan or finish."""

    action: Union[_Response, _Plan] = Field(
        description="If the past steps already answer the user, return _Response. "
        "Otherwise return _Plan with the remaining steps."
    )


_plan_execute_graph = None


def _build_plan_execute_graph():
    """Build the plan-execute StateGraph. Compiled once, cached, reused."""
    llm = make_llm()

    # Planner: turns the user request into an ordered list of atomic steps.
    planner_prompt = ChatPromptTemplate.from_messages(
        [
            (
                "system",
                "For the given user request, produce a concise step-by-step plan. "
                "Each step should be a single concrete action an executor agent "
                "can perform in one turn. Do not skip steps; do not add steps that "
                "aren't needed. The final step should produce the answer.",
            ),
            ("user", "{input}"),
        ]
    )
    planner = (planner_prompt | llm.with_structured_output(_Plan)).with_config(
        {"run_name": "fastapi-langgraph-plan-execute-planner"}
    )

    # Executor: reuses the prebuilt ReAct agent to run each step in turn.
    # By using a sub-agent for execution, the executor handles tool calls
    # (increment, get_counter) without us re-implementing the loop.
    executor = create_react_agent(
        llm,
        tools=TOOLS,
        prompt=SYSTEM_PROMPT,
        name="fastapi-langgraph-plan-execute-executor",
    )

    # Replanner: looks at past steps + remaining plan, decides whether to
    # finish (return _Response) or continue (return _Plan with revised steps).
    replanner_prompt = ChatPromptTemplate.from_messages(
        [
            (
                "system",
                "You are revising an in-progress plan. Given the user's original "
                "request, the original plan, the past steps already executed, "
                "and the remaining plan, either:\n"
                "  - return a _Response if the past steps already answer the request, OR\n"
                "  - return a _Plan with the remaining steps (you may revise them).\n"
                "Prefer _Response when possible — don't repeat work.",
            ),
            (
                "user",
                "Request: {input}\n\nOriginal plan: {plan}\n\n"
                "Past steps:\n{past_steps}\n\nRemaining plan: {remaining}",
            ),
        ]
    )
    replanner = (
        replanner_prompt | llm.with_structured_output(_ReplanAction)
    ).with_config({"run_name": "fastapi-langgraph-plan-execute-replanner"})

    async def plan_step(state: PlanExecuteState):
        plan = await planner.ainvoke({"input": state["input"]})
        return {"plan": plan.steps, "past_steps": []}

    async def execute_step(state: PlanExecuteState):
        if not state["plan"]:
            return {"response": "Plan empty — nothing to execute."}
        task = state["plan"][0]
        # Pass the executor BOTH the original goal and the current step.
        # Without the original goal, the executor sometimes treats steps as
        # read-only (e.g. "get current value") instead of mutating actions
        # (e.g. "call increment()"). With the goal in context, it picks the
        # right tool. Pattern from the LangGraph plan-execute notebook.
        agent_input = {
            "messages": [
                {
                    "role": "user",
                    "content": (
                        f"Overall user request: {state['input']}\n\n"
                        f"Your current sub-step: {task}\n\n"
                        "Use the available tools to actually perform the action. "
                        "Do not just describe — invoke the tool API."
                    ),
                }
            ]
        }
        result = await executor.ainvoke(agent_input)
        last_message = result["messages"][-1]
        result_text = (
            last_message.content
            if hasattr(last_message, "content")
            else str(last_message)
        )
        return {
            "past_steps": [(task, result_text)],
            "plan": state["plan"][1:],
        }

    async def replan_step(state: PlanExecuteState):
        action = await replanner.ainvoke(
            {
                "input": state["input"],
                "plan": state.get("plan", []),
                "past_steps": "\n".join(
                    f"- {step}: {result}" for step, result in state["past_steps"]
                ),
                "remaining": state.get("plan", []),
            }
        )
        if isinstance(action.action, _Response):
            return {"response": action.action.response}
        return {"plan": action.action.steps}

    def should_finish(state: PlanExecuteState):
        # Done when we have a response, or when the plan is empty after a
        # replan (indicating the replanner couldn't produce new steps).
        if state.get("response"):
            return END
        if not state.get("plan"):
            return END
        return "executor"

    workflow = StateGraph(PlanExecuteState)
    workflow.add_node("planner", plan_step)
    workflow.add_node("executor", execute_step)
    workflow.add_node("replanner", replan_step)
    workflow.add_edge(START, "planner")
    workflow.add_edge("planner", "executor")
    workflow.add_edge("executor", "replanner")
    workflow.add_conditional_edges(
        "replanner", should_finish, {END: END, "executor": "executor"}
    )
    return workflow.compile().with_config({"run_name": "fastapi-langgraph-plan-execute"})


def get_plan_execute_graph():
    global _plan_execute_graph
    if _plan_execute_graph is None:
        _plan_execute_graph = _build_plan_execute_graph()
    return _plan_execute_graph


# ---------------------------------------------------------------------------
# Wire-format helpers (shared)
# ---------------------------------------------------------------------------


def _default(obj):
    """Serialize LangChain objects to structured dicts (not repr strings).

    Real LangGraph Cloud emits structured JSON for messages. Without this,
    ToolMessage/AIMessageChunk objects fall back to `str(obj)` which produces
    a Python repr — breaks downstream adapters that expect dict shape.
    """
    if hasattr(obj, "model_dump"):
        try:
            return obj.model_dump()
        except Exception:
            pass
    return str(obj)


def _safe_json(obj) -> str:
    return json.dumps(obj, default=_default)


# ---------------------------------------------------------------------------
# Topology dispatch — each topology has its own state shape, so they have
# their own stream functions rather than one common helper.
# ---------------------------------------------------------------------------


def approval_thread_holds_a_pause(config) -> bool:
    """Does this thread still hold the interrupt a decision would answer? (#399)

    THE DISPATCH CALLS THIS BY NAME ON WHICHEVER MODULE IT ROUTED TO, so arming a
    rung without defining it is an AttributeError on the decisions path and only
    there -- a first message in a gated topology never reaches it. The langchain
    rung has had this since #399; this rung needed it the moment it appeared in
    GATED_TOPOLOGIES, and nothing in the declaration says so.

    THE SAME READER THE EMITTER USES, deliberately, for langchain.py's reason:
    asking a second question a second way is how "the card says pending" and "the
    backend thinks it is pending" come to disagree.

    A LOST THREAD AND A NEVER-RUN THREAD READ IDENTICALLY HERE -- both hold zero
    interrupts -- so this is only half a predicate, and the dispatch supplies the
    other half by asking it only when decisions were actually sent.
    """
    return bool(_pending_approval_events(get_gated_react_graph(), config))


async def stream_chat_react(messages):
    """ReAct topology — prebuilt create_react_agent."""
    # THE DECLARATION DECIDES BOTH ENDS, exactly as on the langchain rung: the
    # dispatch reads GATED_TOPOLOGIES to know whether to demand a policy, and
    # this reads the same constant to know whether to build a gated graph.
    # Deciding independently here is how the two come to disagree.
    gated = "react" in GATED_TOPOLOGIES
    graph = get_gated_react_graph() if gated else get_react_graph()
    # MERGED AT THE CALL SITE, NOT CHOSEN BETWEEN. `langfuse_config()` carries the
    # callbacks and the metadata; the gated path adds `configurable.thread_id`.
    # The first version of this line was a ternary picking one or the other, so
    # every gated turn ran with no callbacks while `/health` went on reporting the
    # backend as traced — an endpoint made to lie by one line. The two dicts share
    # no keys, so a shallow merge is the whole of it, and langchain.py's
    # `_stream_agent_events` has said exactly this since #413; porting this rung's
    # dispatch without reading that far is how the same defect reaches a second rung.
    #
    # WRITTEN IN THE INVOCATION rather than hoisted into a variable, because
    # check-langfuse-wiring reads the call site for `langfuse_config()` and that is
    # not a formality: a hoisted `config=config` is as opaque to someone scanning
    # for untraced paths as it is to the checker.
    thread = approval_thread_config() if gated else {}
    # A RESUME RE-ENTERS THE GRAPH; IT DOES NOT START A TURN. Passing the
    # messages again would append the user's text a second time and run the
    # model afresh, leaving the pending call pending.
    resume = approval_resume_command() if gated else None
    agent_input = resume if resume is not None else {"messages": messages}
    async for event in graph.astream_events(
        agent_input,
        version="v2",
        config={**langfuse_config(), **thread},
    ):
        if _should_emit(event):
            yield f"data: {_safe_json(event)}\n\n"

    # AFTER THE STREAM DRAINS, NOT DURING. An interrupted run ends its event
    # stream normally -- no event names the pause -- so the state can only be
    # asked once the iteration is done.
    if gated:
        for frame in _pending_approval_events(graph, thread):
            yield frame

    yield "data: [DONE]\n\n"


async def stream_chat_plan_execute(messages):
    """Plan-Execute topology — custom StateGraph (planner → executor → replanner)."""
    graph = get_plan_execute_graph()
    user_text = messages[-1]["content"] if messages else ""
    async for event in graph.astream_events(
        {"input": user_text}, version="v2", config=langfuse_config()
    ):
        if _should_emit(event):
            yield f"data: {_safe_json(event)}\n\n"
    yield "data: [DONE]\n\n"


# Public dispatch surface — main.py reads this to route by body.topology.
# WHICH TOPOLOGIES ENFORCE APPROVAL, stated rather than discovered (#332).
#
# ARMED FOR `react`, ON THIS RUNG, ON BOTH PLANES (#332 steps C2/C3).
#
# `plan-execute` is NOT armed and its absence here is a position, not an
# oversight: that topology has not been measured on this rung, and a declaration
# is the wrong place to express a hope. It still relies on the proxy-side
# transform, which withholds the REPORT and not the effect (#256).
#
# BOTH PLANES IN ONE CHANGE, DELIBERATELY. check-run-axes-parity compares this
# constant across the two runtimes since #592, so arming one plane alone is
# correctly red -- the planes would gate different topologies for the same
# request, which is what that check exists to refuse.
GATED_TOPOLOGIES = frozenset({"react"})

TOPOLOGIES = {
    "react": stream_chat_react,
    "plan-execute": stream_chat_plan_execute,
}

# Backward compat: external callers may still reference `stream_chat`.
stream_chat = stream_chat_react


def warmup() -> None:
    """Eager-init so first-request latency and import errors surface at boot.

    Called by main.py's lifespan THROUGH _MODULES, so it disappears with this
    module when `pnpm eject` drops the rung. Do not call it by name from main.
    """
    get_graph()

File: apps/node-backend/src/ai_backends/langgraph.ts
/**
 * LangGraph AI backend for the node runtime — rung 2.
 *
 * RUNG-OWNED FILE, listed under the `langgraph` rung in rungs.json alongside its
 * two Python siblings. Python parity target:
 * apps/fastapi-backend/ai_backends/langgraph.py.
 *
 *   react        → @langchain/langgraph/prebuilt's createReactAgent. The
 *                  prebuilt convenience path, and the common one.
 *   plan-execute → a hand-built StateGraph (planner → executor → replanner with
 *                  a conditional edge back). This is the rung's teaching moment:
 *                  it shows LangGraph's actual abstraction rather than the
 *                  wrapper that hides it. Rung 1 could not express this — its
 *                  plan-execute is a harness AROUND create_agent with no graph
 *                  at all — and that difference is the reason rung 2 exists.
 *
 * WIRE FORMAT: raw `streamEvents` v2 JSON, one event per `data:` line, ending
 * `data: [DONE]`. Same as `langgraph-cli dev` and LangGraph Cloud produce, and
 * the same bytes the Python emits. `langGraphAdapter` normalises it, and its
 * discriminant is `event` — NOT `type`.
 */
import { StateGraph, START, END, Annotation } from "@langchain/langgraph";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";
import { SYSTEM_PROMPT } from "../common/prompt.js";
import { TOOLS } from "../common/tools.js";
import { makeLlm } from "../common/llm.js";
import { runConfig } from "../common/runAxes.js";
import type { ChatMessage } from "./langchain.js";

/** The only events worth putting on the wire. Mirrors `_INTERESTING_EVENTS`. */
const INTERESTING_EVENTS = new Set([
  "on_chat_model_stream",
  "on_tool_start",
  "on_tool_end",
]);

/**
 * Nodes whose `on_chat_model_stream` events are RAW STRUCTURED-OUTPUT JSON.
 *
 * The planner and replanner call the model through `withStructuredOutput`, so
 * their token stream is the serialised schema being generated — `{"steps":["…`
 * — not prose. It is internal data-shape generation and it must not reach the
 * user; the plan surfaces through the executor's own natural-language stream
 * and through the graph state's `response` field.
 *
 * OMITTING THIS FILTER IS THE FAILURE THIS FILE IS MOST LIKELY TO HAVE. It does
 * not produce malformed frames or an adapter error: every frame is a correctly
 * shaped `on_chat_model_stream`, the adapter maps each to a valid AI SDK v6
 * `text-delta`, and the user reads a JSON object where the answer should be. A
 * wire-shape assertion cannot catch it, because the wire shape is right — see
 * langgraph.test.ts, which asserts on CONTENT for exactly that reason.
 */
const STRUCTURED_OUTPUT_NODES = new Set(["planner", "replanner"]);

export interface StreamEvent {
  event: string;
  name?: string;
  run_id?: string;
  data?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export function shouldEmit(event: StreamEvent): boolean {
  if (!INTERESTING_EVENTS.has(event.event)) return false;
  if (event.event === "on_chat_model_stream") {
    const node = (event.metadata?.langgraph_node as string) ?? "";
    if (STRUCTURED_OUTPUT_NODES.has(node)) return false;
  }
  return true;
}

/**
 * Serialise a LangChain object the way the PYTHON plane does, not the way
 * JavaScript does by default.
 *
 * THIS IS A REAL DIVERGENCE AND IT SILENTLY DELETES EVERY TOKEN IF IGNORED.
 * Measured against both sides rather than inferred:
 *
 *   Python  model_dump()      -> {"content": "Hello", "tool_calls": [], …}
 *   JS      JSON.stringify()  -> {"lc":1,"type":"constructor",
 *                                 "id":["langchain_core","messages","AIMessageChunk"],
 *                                 "kwargs":{"content":"hello", …}}
 *
 * `langGraphAdapter` reads `data.chunk.content`. Under the JS default that path
 * is `undefined`, so the adapter emits nothing — and NOTHING ANYWHERE FAILS.
 * The frames are well-formed, the `event` discriminant is correct, the adapter
 * is happy, and the reply is empty. That is the same class as an adapter whose
 * `default` branch passes gibberish through: a green over output that looks
 * like output.
 *
 * So the constructor envelope is unwrapped recursively, leaving the flat shape
 * the fixture shows and the adapter reads.
 */
export function lcSerialize(value: unknown): unknown {
  if (value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) return value.map(lcSerialize);

  const obj = value as Record<string, unknown>;
  // LangChain's serialisation envelope: {lc, type: "constructor", id, kwargs}.
  if (obj.lc === 1 && obj.type === "constructor" && obj.kwargs) {
    return lcSerialize(obj.kwargs);
  }
  // Objects that know how to serialise themselves (messages, chunks) go through
  // toJSON first, which produces the envelope handled above.
  if (typeof (obj as { toJSON?: unknown }).toJSON === "function") {
    return lcSerialize((obj as { toJSON: () => unknown }).toJSON());
  }
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) out[k] = lcSerialize(v);
  return out;
}

/** One `data:` frame carrying a raw streamEvents event. */
export function eventFrame(event: StreamEvent): string {
  return `data: ${JSON.stringify(lcSerialize(event))}\n\n`;
}

export const DONE_FRAME = "data: [DONE]\n\n";

// ---------------------------------------------------------------------------
// Topology 1: ReAct (prebuilt)
// ---------------------------------------------------------------------------

type Graph = { streamEvents: (i: unknown, c: Record<string, unknown>) => AsyncIterable<StreamEvent> };

let reactGraph: Graph | null = null;

export function getReactGraph(): Graph {
  if (reactGraph === null) {
    reactGraph = createReactAgent({
      llm: makeLlm(),
      tools: TOOLS,
      stateModifier: SYSTEM_PROMPT,
      name: "node-langgraph-react",
    }) as unknown as Graph;
  }
  return reactGraph;
}

// ---------------------------------------------------------------------------
// Topology 2: Plan-and-Execute (an explicit StateGraph)
// ---------------------------------------------------------------------------

/**
 * State carried through the plan-execute graph.
 *
 * `input`      the original user request — never mutated.
 * `plan`       ordered list of REMAINING steps. Replaced by the replanner.
 * `pastSteps`  append-only log of [step, result]. Its reducer CONCATENATES,
 *              which is what lets the executor return only its own step and
 *              have LangGraph accumulate the log — the JS equivalent of
 *              Python's `Annotated[List[...], operator.add]`.
 * `response`   terminal field; non-empty means "done, leave the graph".
 */
const PlanExecuteState = Annotation.Root({
  input: Annotation<string>({ reducer: (_, b) => b, default: () => "" }),
  plan: Annotation<string[]>({ reducer: (_, b) => b, default: () => [] }),
  pastSteps: Annotation<Array<[string, string]>>({
    reducer: (a, b) => a.concat(b),
    default: () => [],
  }),
  response: Annotation<string>({ reducer: (_, b) => b, default: () => "" }),
});

/** Planner output. JSON Schema, not zod — see common/tools.ts for why. */
const PLAN_SCHEMA = {
  type: "object",
  properties: {
    steps: {
      type: "array",
      items: { type: "string" },
      description:
        "Ordered, atomic steps to accomplish the user's request. Each step " +
        "should be specific enough that an executor agent can act on it in a " +
        "single turn.",
    },
  },
  required: ["steps"],
} as const;

/**
 * Replanner output — either finish with a response, or continue with a plan.
 *
 * Python models this as `Union[_Response, _Plan]`. A JSON Schema union would be
 * `oneOf`, which providers support unevenly through tool-calling, so this is one
 * object with both fields optional and `response` taking precedence. The
 * BEHAVIOUR is identical — the branch below reads exactly what Python's
 * `isinstance(action.action, _Response)` reads — and the difference is recorded
 * rather than smoothed over.
 */
const REPLAN_SCHEMA = {
  type: "object",
  properties: {
    response: {
      type: "string",
      description:
        "Set this ONLY if the past steps already answer the user's request. " +
        "Prefer this when possible — do not repeat work.",
    },
    steps: {
      type: "array",
      items: { type: "string" },
      description:
        "The remaining steps, possibly revised. Set this only if more work " +
        "is genuinely needed.",
    },
  },
} as const;

const PLANNER_SYSTEM =
  "For the given user request, produce a concise step-by-step plan. Each step " +
  "should be a single concrete action an executor agent can perform in one " +
  "turn. Do not skip steps; do not add steps that aren't needed. The final " +
  "step should produce the answer.";

const REPLANNER_SYSTEM =
  "You are revising an in-progress plan. Given the user's original request, " +
  "the past steps already executed, and the remaining plan, either set " +
  "`response` if the past steps already answer the request, or set `steps` to " +
  "the remaining steps (you may revise them). Prefer `response` when possible " +
  "— don't repeat work.";

let planExecuteGraph: Graph | null = null;

export function buildPlanExecuteGraph(model?: BaseChatModel): Graph {
  const llm = model ?? makeLlm();

  // NAMED `planner` AND `replanner` FOR A REASON BEYOND READABILITY: those two
  // names are the keys STRUCTURED_OUTPUT_NODES matches on, via
  // metadata.langgraph_node. Rename a node and its raw JSON starts reaching the
  // user — which is why langgraph.test.ts asserts the suppression by node name
  // rather than by position.
  const planner = llm.withStructuredOutput(PLAN_SCHEMA, { name: "Plan" });
  const replanner = llm.withStructuredOutput(REPLAN_SCHEMA, {
    name: "ReplanAction",
  });

  // The executor reuses the prebuilt ReAct agent, so tool calls (increment,
  // get_counter) are handled without re-implementing the loop — the same choice
  // the Python makes, and the reason the two rungs share a tool surface.
  const executor = createReactAgent({
    llm,
    tools: TOOLS,
    stateModifier: SYSTEM_PROMPT,
    name: "node-langgraph-plan-execute-executor",
  });

  const workflow = new StateGraph(PlanExecuteState)
    .addNode("planner", async (state) => {
      const plan = (await planner.invoke([
        { role: "system", content: PLANNER_SYSTEM },
        { role: "user", content: state.input },
      ])) as { steps?: string[] };
      return { plan: plan.steps ?? [], pastSteps: [] };
    })
    .addNode("executor", async (state) => {
      if (state.plan.length === 0) {
        return { response: "Plan empty — nothing to execute." };
      }
      const task = state.plan[0];
      // BOTH THE GOAL AND THE STEP. Without the original goal the executor
      // sometimes treats a step as read-only ("get current value") instead of
      // the mutating action it is ("call increment()"). Pattern taken from the
      // LangGraph plan-execute notebook, and the Python carries the same note.
      const result = (await executor.invoke({
        messages: [
          {
            role: "user",
            content:
              `Overall user request: ${state.input}\n\n` +
              `Your current sub-step: ${task}\n\n` +
              "Use the available tools to actually perform the action. " +
              "Do not just describe — invoke the tool API.",
          },
        ],
      })) as { messages: Array<{ content?: unknown }> };
      const last = result.messages[result.messages.length - 1];
      const text =
        typeof last?.content === "string" ? last.content : String(last?.content ?? "");
      return {
        pastSteps: [[task, text]] as Array<[string, string]>,
        plan: state.plan.slice(1),
      };
    })
    .addNode("replanner", async (state) => {
      const action = (await replanner.invoke([
        { role: "system", content: REPLANNER_SYSTEM },
        {
          role: "user",
          content:
            `Request: ${state.input}\n\n` +
            `Past steps:\n${state.pastSteps
              .map(([step, result]) => `- ${step}: ${result}`)
              .join("\n")}\n\n` +
            `Remaining plan: ${JSON.stringify(state.plan)}`,
        },
      ])) as { response?: string; steps?: string[] };
      // `response` wins, mirroring Python's isinstance(_Response) branch.
      if (action.response) return { response: action.response };
      return { plan: action.steps ?? [] };
    })
    .addEdge(START, "planner")
    .addEdge("planner", "executor")
    .addEdge("executor", "replanner")
    // THE CONDITIONAL EDGE IS THE POINT OF THIS TOPOLOGY. Finish when there is
    // a response, or when the plan is empty after a replan (the replanner could
    // not produce new steps); otherwise loop back into the executor.
    .addConditionalEdges(
      "replanner",
      (state) => {
        if (state.response) return END;
        if (state.plan.length === 0) return END;
        return "executor";
      },
      { [END]: END, executor: "executor" }
    );

  return workflow.compile() as unknown as Graph;
}

export function getPlanExecuteGraph(): Graph {
  if (planExecuteGraph === null) planExecuteGraph = buildPlanExecuteGraph();
  return planExecuteGraph;
}

// ---------------------------------------------------------------------------
// Topology dispatch — each topology has its own state shape, so each has its
// own stream function rather than one shared helper. Same as the Python.
// ---------------------------------------------------------------------------

async function* streamGraph(
  graph: Graph,
  input: unknown
): AsyncGenerator<string> {
  for await (const event of graph.streamEvents(input, {
    version: "v2",
    ...runConfig(),
  })) {
    if (shouldEmit(event)) yield eventFrame(event);
  }
  yield DONE_FRAME;
}

export async function* streamChatReact(
  messages: ChatMessage[]
): AsyncGenerator<string> {
  yield* streamGraph(getReactGraph(), { messages });
}

export async function* streamChatPlanExecute(
  messages: ChatMessage[]
): AsyncGenerator<string> {
  const userText = messages.length > 0 ? messages[messages.length - 1].content : "";
  yield* streamGraph(getPlanExecuteGraph(), { input: userText });
}

export const TOPOLOGIES: Record<
  string,
  (messages: ChatMessage[]) => AsyncGenerator<string>
> = {
  react: streamChatReact,
  "plan-execute": streamChatPlanExecute,
};

export function warmup(): void {
  getReactGraph();
  getPlanExecuteGraph();
}
